log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/instant/InstantPatternDynamicFormatter.java

```diff
@@ -806,26 +674,114 @@ private static String removePadding(final String content) {
+            // Skip digits beyond the requested precision
+            while (fractionalDigits < valuePrecision) {
+                valuePrecision--;
+                value = value / 10L;
+            }
+            // Print the digits
+            while (0 < valuePrecision--) {
+                buffer.setCharAt(offset + valuePrecision, (char) ('0' + value % 10L));
+                value = value / 10L;
+            }
+        }
+
+        @Override
+        InstantPatternFormatter createFormatter(Locale locale, TimeZone timeZone) {
+            if (!printSeconds) {
+                return new AbstractFormatter(pattern, locale, timeZone, precision) {
+                    @Override
+                    public void formatTo(StringBuilder buffer, Instant instant) {
+                        buffer.append(separator);
+                        formatFractionalDigits(buffer, instant);
+                    }
+                };
+            }
+            if (fractionalDigits == 0) {
+                return new AbstractFormatter(pattern, locale, timeZone, precision) {
+                    @Override
+                    public void formatTo(StringBuilder buffer, Instant instant) {
+                        formatSeconds(buffer, instant);
+                        buffer.append(separator);
+                    }
+                };
+            }
+            return new AbstractFormatter(pattern, locale, timeZone, precision) {
+                @Override
+                public void formatTo(StringBuilder buffer, Instant instant) {
+                    formatSeconds(buffer, instant);
+                    buffer.append(separator);
+                    formatFractionalDigits(buffer, instant);
+                }
+            };
+        }
+
+        @Override
+        @Nullable
+        PatternFormatterFactory tryMerge(PatternFormatterFactory other, ChronoUnit thresholdPrecision) {
+            // If we don't have a fractional part, we can merge a literal separator
+            if (other instanceof StaticPatternFormatterFactory) {
+                StaticPatternFormatterFactory staticOther = (StaticPatternFormatterFactory) other;
+                if (fractionalDigits == 0) {
+                    return new SecondPatternFormatterFactory(
+                            printSeconds, this.separator + staticOther.literal, fractionalDigits);
+                }
+            }
+            // We can always append more fractional digits
+            if (other instanceof SecondPatternFormatterFactory) {
+                SecondPatternFormatterFactory secondOther = (SecondPatternFormatterFactory) other;
+                if (!secondOther.printSeconds && secondOther.separator.isEmpty()) {
+                    return new SecondPatternFormatterFactory(
+                            printSeconds, this.separator, this.fractionalDigits + secondOther.fractionalDigits);
+                }
+            }
+            return super.tryMerge(other, thresholdPrecision);
```

Comment:
One less cognitive hop:
```suggestion
            return null;
```